middleware/reverse/reverse.go

```diff
@@ -0,0 +1,109 @@
+package reverse
+
+import (
+	"net"
+
+	"github.com/miekg/coredns/middleware"
+	"golang.org/x/net/context"
+	"github.com/miekg/dns"
+	"github.com/miekg/coredns/request"
+	"github.com/miekg/coredns/middleware/pkg/dnsutil"
+)
+
+// Reverse provides dynamic reverse dns and the related forward rr
+type Reverse struct {
+	Next     middleware.Handler
+	Networks Networks
+}
+
+// ServeDNS implements the middleware.Handler interface.
+func (reverse Reverse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	var rr dns.RR
+
+	state := request.Request{W: w, Req: r}
+	m := new(dns.Msg)
+	m.SetReply(r)
+	m.Authoritative, m.RecursionAvailable, m.Compress = true, true, true
+
+	// Reverse Lookup
+	if state.QType() == dns.TypePTR {
```

Comment:
Just group this under the switch you start below, then you also can drop the indenting one level and remove the `else` when doing the forward lookup stuff.